addons/website/static/src/js/editor/snippets.options.js

```diff
@@ -1096,6 +1108,43 @@ options.registry.layout_column = options.Class.extend({
     },
 });
 
+options.registry.ZeroColumns = options.registry.layout_column.extend({
+    /**
+     * @override
+     */
+     _computeWidgetState(methodName, params) {
+         if (methodName === 'colCount') {
+             return this.$('> .row:not(.o_we_zero_cols_content)').children().length;
+         }
+         return this._super(...arguments);
+     },
+
+    //--------------------------------------------------------------------------
+    // Options
+    //--------------------------------------------------------------------------
+
+    /**
+     * Changes the number of columns.
+     *
+     * @private
+     */
+    async colCount(previewMode, widgetValue, params) {
+        const count = parseInt(widgetValue);
+        let $row = this.$('> .row:not(.o_we_zero_cols_content)');
+        if (!$row.length) {
+            $row = $('<div class="row"><div class="col-lg-12"></div></div>');
+            $row.children().first().append(this.$target.children());
+            this.$target.append($row);
+        }
+        const proxy = createPropertyProxy(this, {$target: $row});
+        await proxy._updateColumnCount((count || 1) - $row.children().length);
```

Comment:
I was initially doing it this way, but I was having to treat the zero-case specially all throughout the option. This way has the advantage that it keeps all the zero-case handling code in one place, since it is, in fact, a special case. It also involved reworking a lot of the logic of the original option because the `$target` can no longer be the row since it can't be assumed to be present. It also had the problem that I had to manually manage the UI of the 0 button of the select since it's not a top-level item in the option and can't use `_computeWidgetVisibility`.

For all those reasons I thought it would be cleaner to make a separate option that inherits from the original, at the (IMO lesser) cost of duplicating a bit of XML. I had initially done as you're suggesting and I found it very messy and that it added a lot of complexity to the code, but to be fair I was very tired that day and I would probably be able to get it a lot cleaner now that my brain is back to normal. The only remaining problem would be manual UI management for the 0 item in the select, which may be preferable to the XML duplication that I have in this version.

As far as using a proxy, it was is in the same spirit as the `data-apply-to` attribute: since the method is async, I didn't want to have the same problem as we have with `_super` where the `$target` would be incorrect in asynchronous code.

What I didn't realize at the time is that since option methods are executed in a mutex, and they cannot run in parallel, we probably didn't need a proxy in the first place and we could just patch `$target` before the call and restore it after (and in this case we could do the same). The only exception would be if someone had code that ran *after* the option method returned, for example in a `setTimeout` or any other callback-based function that is not awaited within the option method, but that seems like it's not something we want to allow since the whole point of awaiting option methods is to ensure they finished their work.

So here's what I would like to do if that's ok with you:
* 1 commit for making `onCloneSnippet` and `onRemoveSnippet` take a `onSuccess` or `callback` param so that they can be awaited
* 1 commit to remove the use of property proxies since they're basically there to support a use case that we explicitly don't support (multiple option methods running in parallel)
* 1 commit to add the zero column capability to the existing layout_column option

Reply:
- Commit 1: yes :+1: 
- Commit 2: Are you sure it was the only reason for using a proxy? Because when you did it, I did suggest to only change `$target` for the duration of the option method call and I don't think you convinced me just with the option-in-parallel argument... :/ but maybe I forgot, the mutex did appear for 13.2 not 13.1 so it is possible. (Not: in fact, if `apply-to` matches multiple elements, the options could still run in parallel for each element, justifying the use of the proxy but maybe it is a bad reason).
- Commit 3: Yes, I don't think it will be messy to go with an unique option Class handling both cases. The `$target` becomes the `.container` and it is just about finding/creating the row and leaving the old code work on that row. About the XML duplication, it is not a problem, you can leave it as a duplication for two different option instances, using the same option Class. So:
```xml
<div data-js="layout_column" data-selector="XXX" data-target=".container:not(.somethingSpecific)">
    <we-select>... choices without 0 available ...</we-select>
</div>
<div data-js="layout_column" data-selector="XXX" data-target=".container.somethingSpecific" ...>
    <we-select>... choices with 0 available ...</we-select>
</div>
```
and you can just avoid duplication of the XML with some t-set/t-raw/t-call magic. But we do need a way to have a `computeVisibility` of internal widgets of a select, but this can be a workaround for now.

![](https://www.eduscapes.com/history/ancient/romanroll.jpg)

Reply:
> Note: in fact, if apply-to matches multiple elements, the options could still run in parallel for each element, justifying the use of the proxy but maybe it is a bad reason

They could (and at the moment, they do), but considering the initial idea was to mimic as closely as possible the behaviour of an option method on a different target, it seems like a bad idea in retrospect. Also, looking at the code right now there's a nasty bug waiting to happen:

```javascript
if (params.applyTo) {
    const $subTargets = this.$(params.applyTo);
    const proms = _.each($subTargets, subTargetEl => {
        const proxy = createPropertyProxy(this, {$target: $(subTargetEl)});
        return this[methodName].call(proxy, previewMode, widgetValue, params);
    });
    await Promise.all(proms);
}
```

It's using `each` instead of `map`, so the promises are never awaited.

I'm pretty sure my initial intention was to avoid replicating the `this._super` behaviour where you have to cache it before any async work and I just didn't take into account that the behaviour was substantially different. If you can think of or recall any other reason why we might want to use a proxy, I'm perfectly fine with leaving it in, I'd just rather not leave it there if it's not needed as it makes it adds a level of indirection when reasoning about options.

Reply:
Ok, then let's remove the proxies in master. And you can make a PR to fix that each/map fail in the oldest stable version (probably 13.2 ?), thanks :)

Reply:
each/map: #44959